### PR TITLE
C++: Promote `cpp/guarded-free` out of experimental

### PR DESCRIPTION
**Pull request by jketema**
#### All query authors

- [x] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [x] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [x] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [x] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [x] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

**Review by geoffw0 — COMMENTED** [review]
Changes LGTM.  A sample of the new DCA results LGTM.

We could replace `FreeCall` with `DeallocationExpr` and `fc.getArgument(0)` with `fc.getFreedExpr()`, to add results for all kinds of `free` variants including, I think, `delete`.  This is assuming we're confident all of these variants are in fact safe on `NULL`.  It probably makes sense to promote the query as it is and consider this as a follow-up step afterwards?

Needs a docs review.

**Comment by mchammer01** [comment]
I'll review this on behalf of Docs.

File: cpp/ql/src/Best Practices/GuardedFree.qhelp [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
+
 <recommendation>
-<p>A function call to <code>free</code> should not depend upon the value of its argument. Delete the <code>if</code> condition preceeding a function call to <code>free</code> when its only purpose is to check the value of the pointer to be freed.</p>
+<p>A function call to <code>free</code> should not depend upon the value of its argument. Delete the condition preceding a function call to <code>free</code> when its only purpose is to check the value of the pointer to be freed.</p>
 </recommendation>
+
 <example>
 <sample src = "GuardedFree.cpp" />
+
+<p>In this example the condition checking the value of <code>foo</code> can be deleted.</p>
```
Comment (mchammer01) [inline cpp/ql/src/Best Practices/GuardedFree.qhelp:17]: Nit 
```suggestion
<p>In this example, the condition checking the value of <code>foo</code> can be deleted.</p>
```

File: cpp/ql/src/Best Practices/GuardedFree.qhelp [hunk 2]
```diff
@@ -1,18 +1,28 @@
-<!DOCTYPE qhelp SYSTEM "qhelp.dtd">
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
 <qhelp>
+
 <overview>
 <p>The <code>free</code> function, which deallocates heap memory, may accept a NULL pointer and take no action. Therefore, it is unnecessary to check its argument for the value of NULL before a function call to <code>free</code>. As such, these guards may hinder performance and readability.</p>
```
Comment (mchammer01) [inline cpp/ql/src/Best Practices/GuardedFree.qhelp:7]: Slightly rewording as the pronoun is currently before the noun it applies to.

```suggestion
<p>The <code>free</code> function, which deallocates heap memory, may accept a NULL pointer and take no action. Therefore, it is unnecessary to check the argument for the value of NULL before a function call to <code>free</code>. As such, these guards may hinder performance and readability.</p>
```

**Review by mchammer01 — APPROVED** [review]
LGTM ✨ 
Approving this but left a few comments and suggestions following editorial review.

**Comment by jketema** [comment]
@geoffw0 Would you mind re-approving so we can merge this?

**Review by geoffw0 — APPROVED** [review]
:+1:

**Pull request merged**